File: vane-core/src/main/java/org/oddlama/vane/core/resourcepack/ResourcePackDevServer.java
package org.oddlama.vane.core.resourcepack;

import com.google.common.hash.Hashing;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;

public class ResourcePackDevServer implements HttpHandler {

    private final ResourcePackDistributor resourcePackDistributor;
    private final File file;

    public ResourcePackDevServer(ResourcePackDistributor resourcePackDistributor, File file) {
        this.resourcePackDistributor = resourcePackDistributor;
        this.file = file;
    }

    @SuppressWarnings({ "deprecation", "UnstableApiUsage" })
    public void serve() {
        try {
            final HttpServer httpServer = HttpServer.create(new InetSocketAddress(9000), 0);
            var hash = com.google.common.io.Files.asByteSource(this.file).hash(Hashing.sha1());
            resourcePackDistributor.packSha1 = hash.toString();
            resourcePackDistributor.packUrl = "http://localhost:9000/vane-resource-pack.zip";

            httpServer.createContext("/", this);
            httpServer.setExecutor(null);
            httpServer.start();
        } catch (IOException e) {
            resourcePackDistributor.getModule().log.log(
                java.util.logging.Level.SEVERE,
                "Failed to start resource pack dev server",
                e
            );
        }
    }

    public void handle(HttpExchange he) throws IOException {
        String method = he.getRequestMethod();
        if (!("HEAD".equals(method) || "GET".equals(method))) {
            he.sendResponseHeaders(501, -1);
            return;
        }

        FileInputStream fis;
        try {
            fis = new FileInputStream(file);
        } catch (FileNotFoundException e) {
            he.sendResponseHeaders(404, -1);
            return;
        }

        he.getResponseHeaders().set("Content-Type", "application/zip");
        if ("GET".equals(method)) {
            he.sendResponseHeaders(200, file.length());
            OutputStream os = he.getResponseBody();
            fis.transferTo(os);
            os.close();
        } else {
            he.sendResponseHeaders(200, -1);
        }
        fis.close();
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/resourcepack/ResourcePackDevServer.kt
package org.oddlama.vane.core.resourcepack

import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpHandler
import com.sun.net.httpserver.HttpServer
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.net.InetSocketAddress
import java.security.MessageDigest
import java.util.logging.Level

/**
 * Hosts a local HTTP server that serves a generated resource pack during development.
 *
 * @param resourcePackDistributor the distributor whose URL and SHA-1 are kept in sync.
 * @param file the resource pack zip file to serve.
 */
class ResourcePackDevServer(private val resourcePackDistributor: ResourcePackDistributor, private val file: File) :
    HttpHandler {
    /**
     * The running HTTP server instance, if active.
     */
    private var httpServer: HttpServer? = null

    /**
     * Stops the HTTP server if it is running.
     */
    fun stop() {
        httpServer?.stop(0)
        httpServer = null
    }

    /**
     * Starts the HTTP server and updates the distributor URL and SHA-1 hash.
     */
    fun serve() {
        try {
            val server = HttpServer.create(InetSocketAddress(9000), 0)
            val hashBytes = MessageDigest.getInstance("SHA-1").digest(file.readBytes())
            resourcePackDistributor.packSha1 = hashBytes.joinToString("") { "%02x".format(it) }
            resourcePackDistributor.packUrl = "http://localhost:9000/vane-resource-pack.zip"

            server.apply {
                createContext("/", this@ResourcePackDevServer)
                executor = null
                start()
            }
            httpServer = server
        } catch (e: IOException) {
            resourcePackDistributor.module?.log?.log(
                Level.SEVERE,
                "Failed to start resource pack dev server",
                e
            )
        }
    }

    /**
     * Handles HTTP requests for the resource pack file.
     *
     * Supports `HEAD` and `GET`; returns `404` when the zip file is missing.
     *
     * @param he the incoming HTTP exchange.
     * @throws IOException if request handling fails.
     */
    @Throws(IOException::class)
    override fun handle(he: HttpExchange) {
        val method = he.requestMethod
        if (method !in setOf("HEAD", "GET")) {
            he.sendResponseHeaders(501, -1)
            return
        }

        val fis = try {
            file.inputStream()
        } catch (_: FileNotFoundException) {
            he.sendResponseHeaders(404, -1)
            return
        }

        he.responseHeaders.set("Content-Type", "application/zip")
        fis.use {
            if (method == "GET") {
                he.sendResponseHeaders(200, file.length())
                he.responseBody.use { os -> it.transferTo(os) }
            } else {
                he.sendResponseHeaders(200, -1)
            }
        }
    }
}
